### ISSUE TITLE
Error when retraining model

### ISSUE THREAD

**Issue by jwood803** [dotnet/docs#13898]
When calling the `Fit` method with the original model parameters to retrain a model, I get the below error message. I'm not quite sure if I did something incorrect to cause this or if it is from something else.

```
Optimizer unable to proceed with loss function yielding Infinity
```

I did have to use the `LinearModelParameters` when casting the `Model` parameter on the loaded model file. This was because, if casting as `LinearRegressionModelParameters`, it wouldn't cast and would return `null`.

```
var originalModelParams =
    ((ISingleFeaturePredictionTransformer&lt;object&gt;)trainerModel).Model as LinearModelParameters;
```

Here's the [full code](https://github.com/jwood803/MLNetExamples/blob/master/MLNetExamples/AzureFunction/HousingRetrain.cs) if needed.

Thanks!

[AB#1579020](https://dev.azure.com/mseng/677da0fb-b067-4f77-b89b-f32c12bb8617/_workitems/edit/1579020)

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 0366cc46-bc07-44fd-72d9-d37b40bd5eca
* Version Independent ID: 5c20d041-f946-d7c8-b0b5-7bee96e90282
* Content: [Re-train a model - ML.NET](https://docs.microsoft.com/en-us/dotnet/machine-learning/how-to-guides/retrain-model-ml-net#feedback)
* Content Source: [docs/machine-learning/how-to-guides/retrain-model-ml-net.md](https://github.com/dotnet/docs/blob/master/docs/machine-learning/how-to-guides/retrain-model-ml-net.md)
* Product: **dotnet-ml**
* GitHub Login: @luisquintanilla
* Microsoft Alias: **luquinta**

**Comment by luisquintanilla** [comment]
Hi @jwood803 ,

Thanks for your question. The model parameter type is dependent on the model that is being trained. Try changing `LinearModelParameters` to `PoissonRegressionModelParameters`. A good way to tell is by inspecting the [API documentation](https://docs.microsoft.com/en-us/dotnet/api/microsoft.ml.trainers.lbfgspoissonregressiontrainer?view=ml-dotnet). 

This is the signature for the `LbfgsPoissonRegressionTrainer`. Make special note of the model parameter type which is `PoissonRegressionModelParameters`:

public sealed class LbfgsPoissonRegressionTrainer : Microsoft.ML.Trainers.LbfgsTrainerBase<Microsoft.ML.Trainers.LbfgsPoissonRegressionTrainer.Options,Microsoft.ML.Data.RegressionPredictionTransformer<__Microsoft.ML.Trainers.PoissonRegressionModelParameters__>,__Microsoft.ML.Trainers.PoissonRegressionModelParameters__>

Hope this solves the problem.

See this full working sample:

```csharp
using System;
using System.Linq;
using Microsoft.ML;
using Microsoft.ML.Data;
using Microsoft.ML.Trainers;

namespace RetrainSample
{
    class HousingData
    {
        [LoadColumn(0)]
        public float Size { get; set; }

        [LoadColumn(1), VectorType(3)]
        public float[] HistoricalPrices { get; set; }

        [LoadColumn(2), ColumnName("Label")]
        public float CurrentPrice { get; set; }
    }

    class Program
    {
        static void Main(string[] args)
        {
            MLContext mlContext = new MLContext();

            Train(mlContext);

            Retrain(mlContext);

            Console.ReadKey();
        }

        static void Train(MLContext mlContext)
        {
            HousingData[] housingData = new HousingData[] {
                new HousingData
                {
                    Size = 600f,
                    HistoricalPrices = new float[] { 100000f ,125000f ,122000f },
                    CurrentPrice = 170000f
                },
                new HousingData
                {
                    Size = 1000f,
                    HistoricalPrices = new float[] { 200000f, 250000f, 230000f },
                    CurrentPrice = 225000f
                },
                new HousingData
                {
                    Size = 1000f,
                    HistoricalPrices = new float[] { 126000f, 130000f, 200000f },
                    CurrentPrice = 195000f
                }
            };

            var data = mlContext.Data.LoadFromEnumerable(housingData);

            IEstimator<ITransformer> pipelineEstimator =
                mlContext.Transforms.Concatenate("Features", new string[] { "Size", "HistoricalPrices" })
                .Append(mlContext.Transforms.NormalizeMinMax("Features"));

            ITransformer dataPrepTransformer = pipelineEstimator.Fit(data);

            var regressionEstimator = mlContext.Regression.Trainers.LbfgsPoissonRegression();

            IDataView transformedData = dataPrepTransformer.Transform(data);

            ITransformer trainedModel = regressionEstimator.Fit(transformedData);

            mlContext.Model.Save(dataPrepTransformer, data.Schema, "dataprep.zip");
            mlContext.Model.Save(trainedModel, transformedData.Schema, "model.zip");
        }

        static void Retrain(MLContext mlContext)
        {
            // Load data preparation pipeline
            ITransformer dataPrepPipeline = mlContext.Model.Load("dataprep.zip", out DataViewSchema dataPrepPipelineSchema);

            // Load trained model
            ITransformer trainedModel = mlContext.Model.Load("model.zip", out DataViewSchema modelSchema);

            PoissonRegressionModelParameters originalModelParameters = ((ISingleFeaturePredictionTransformer<object>)trainedModel).Model as PoissonRegressionModelParameters;

            HousingData[] housingData = new HousingData[]
            {
                new HousingData
                {
                    Size = 850f,
                    HistoricalPrices = new float[] { 150000f,175000f,210000f },
                    CurrentPrice = 205000f
                },
                new HousingData
                {
                    Size = 900f,
                    HistoricalPrices = new float[] { 155000f, 190000f, 220000f },
                    CurrentPrice = 210000f
                },
                new HousingData
                {
                    Size = 550f,
                    HistoricalPrices = new float[] { 99000f, 98000f, 130000f },
                    CurrentPrice = 180000f
                }
            };

            //Load New Data
            IDataView newData = mlContext.Data.LoadFromEnumerable(housingData);

            // Preprocess Data
            IDataView transformedNewData = dataPrepPipeline.Transform(newData);

            // Retrain model
            RegressionPredictionTransformer<PoissonRegressionModelParameters> retrainedModel =
                mlContext.Regression.Trainers.LbfgsPoissonRegression()
                    .Fit(transformedNewData, originalModelParameters);

            // Extract Model Parameters of re-trained model
            PoissonRegressionModelParameters retrainedModelParameters = retrainedModel.Model as PoissonRegressionModelParameters;

            // Inspect Change in Weights
            var weightDiffs =
                originalModelParameters.Weights.Zip(
                    retrainedModelParameters.Weights, (original, retrained) => original - retrained).ToArray();

            Console.WriteLine("Original | Retrained | Difference");
            for (int i = 0; i < weightDiffs.Count(); i++)
            {
                Console.WriteLine($"{originalModelParameters.Weights[i]} | {retrainedModelParameters.Weights[i]} | {weightDiffs[i]}");
            }
        }
    }
}
```

**Comment by jwood803** [comment]
Hey, @luisquintanilla. Thank you for the detailed reply! Casting to `PoissonRegressionModelParameters` was able to get the correct object. Unfortunately, during the below code to the `Fit` method, I still get the "Optimizer unable to proceed with loss function yielding Infinity" error. 

```
var retrainedModel = context.Regression.Trainers.LbfgsPoissonRegression()
                .Fit(newDataTransformed, originalModelParams);
```

Thanks for the help!

**Comment by luisquintanilla** [comment]
@jwood803 If you try to retrain the model locally (not using Azure Functions and Storage Blob), do you still get the same error?

**Comment by jwood803** [comment]
Hey, @luisquintanilla. I just tried with a console app and I do still get the same exception.

Here's the [code](https://github.com/jwood803/MLNetExamples/tree/master/MLNetExamples/RetrainModel), if you need. Let me know if I can help with anything. :)

**Comment by luisquintanilla** [comment]
Thanks for this @jwood803. I would suggest in this case raising an issue in the dotnet/machinelearning repo.

**Comment by jwood803** [comment]
Thanks @luisquintanilla! I appreciate all the help!

I just logged [issue #4230](https://github.com/dotnet/machinelearning/issues/4230).
